Real ranged attacks

lipk opened the pull request:
Reintroduce real ranged attacks to Wesnoth.

[Forum thread](https://forums.wesnoth.org/viewtopic.php?f=12&p=654394#p654394)

What the PR does:
* Allows targeting enemies from multiple tiles away
* Applies distance-based penalties to hit chance and damage of a weapon
* Exposes penalties and ranges to WML and Lua
* Displays information about ranged attacks on the unit's help page

Necessary actions after this PR is merged:
- [ ] Update UnitTypeWML docs with new weapon attributes
- [ ] Update FilterWML docs with new weapon filters
- [ ] Update EffectWML docs with new weapon modifiers

I believe this is a viable minimal implementation and can be merged back to allow content creators to experiment. Nonetheless, there are a couple more things that need to be done before this feature can be deemed ready.

TODO:
* Proper animation support for multihex attacks
* Make the AI aware of real ranged attacks

Things that may need further consideration:
* How the new information is displayed on the help page 
* Visual aid to indicate which units will be in range when moving to a certain tile
* Different attack marker for ranged attacks

Pentarctagon commented:
AI would be @mattsc .  What's meant by "Proper animation support for multihex attacks"?

lipk commented:
Making the missile fly all the way to the target. Right now it vanishes at the first hex.

gfgtdf commented:
how does it look like on the ui? meaning in particular how is the arrow drawn? straight to the target ot with edges in betwwen as in pathfinding?

gfgtdf commented on src/mouse_events.cpp:
```diff
@@ -1405,19 +1406,17 @@ std::set<map_location> mouse_handler::get_adj_enemies(const map_location& loc, i
-
-	for(const map_location& loc : adj) {
+	// Check each tile in the unit's attack range
+	auto range = get_tiles_in_range(u->get_location(), u->min_attack_range(), u->max_attack_range());
```
you added an assertion on about hte parameters on get_tiles_in_range but i see no code here that makes sure the assertion is true.

gfgtdf replied:
furthermore it might happen that your units has multiple attacks, one for long range and one for short range, and on first sight this looks like it will make all available.

gfgtdf replied:
> but i see no code here that makes sure the assertion is true.

Nevermind it seems to be slot in the setter in attack min/max range

lipk replied:
It's in attack_type.cpp. min_range is always initialized to at least 1 and max_range to at least min_range.

However, I now realize that later on the >0 constraint is not enforced when modifying the values. I'll fix that.

lipk replied:
> furthermore it might happen that your units has multiple attacks, one for long range and one for short range, and on first sight this looks like it will make all available.

This part of the code only controls which units can be targeted. So any enemy that you can reach with at least one weapon should be highlighted. Later, the attack dialog will filter out the weapons without proper range (this was already there).

CelticMinstrel commented:
I notice that `get_tiles_in_range` seems to be using a new algorithm, and I wonder if it differs in unexpected ways from other similar algorithms in the engine, in particular the use of `radius` in a standard location filter.

For the damage and hit chance penalties, I'd drop the `_formula` suffix and use `type_formula` to store the formulas in code. In terms of syntax in the WML, `typed_formula` would mean that unless it's a constant value, you need to enclose the entire formula in parentheses; but it avoids parsing and executing a formula when the value is constant. This also means there'd be (as far as I can tell) no need for two separate damage penalty attributes.

Feels like weapon filter and modifications might be missing a few things, though some of them may not have an obvious implementation — filter on hit chance and damage penalties, increase min range, increase (or even transform) hit chance and damage penalties. Perhaps they can wait for later, though (other than increasing min range which seems trivial?).

lipk commented:
@gfgtdf Only the arrow is visible
![wesnoth-ranged-attack](https://user-images.githubusercontent.com/3096858/79695675-82c73080-8278-11ea-8fac-ca2f0d6ec4d6.png)

gfgtdf commented:
> Only the arrow is visible

and the start of the arrow on top the attacker units.

I really think it be better if we would show a full arrow from the attacker to the defender. ideally a 'diagonal' one and not a pathfind-like arrow.

mattsc commented:
> AI would be @mattsc .

Actually, not really.  I know a bit about AI behavior, not so much AI (code) mechanics.  And assuming that from the point of view of the AI this is just another attack to choose from, then this is a mechanics question.  Besides that, I neither have the time not the energy to deal with this in the foreseeable future (and for that same reason have not read up on how exactly this is supposed to work).

lipk commented:
> and the start of the arrow on top the attacker units.
Ah. I haven't noticed that :)

> I really think it be better if we would show a full arrow from the attacker to the defender. ideally a 'diagonal' one and not a pathfind-like arrow.
I'd put that way down the bottom of the todo list. It's a lot of work for a small eyecandy. I think adding a non-directional replacement for the arrow (crosshairs or something similar) should be good enough, and even the arrow doesn't look that terrible.

lipk commented:
> I notice that `get_tiles_in_range` seems to be using a new algorithm, and I wonder if it differs in unexpected ways from other similar algorithms in the engine, in particular the use of `radius` in a standard location filter.

I'm not sure what do you mean by 'unexpected ways'. It solves a rigorously defined problem. Unless I screwed something up in the implementation, it should give correct results 100% of the time.

I looked up the location filter. It eventually calls get_tiles_in_radius, which uses a different (and I think slightly less efficient) implementation than the new function. I'm a bit reluctant to ditch a faster implementation for slower one, but probably the correct way to proceed is to refactor my code to use get_tiles_in_radius and then optimize that function later.

> For the damage and hit chance penalties, I'd drop the `_formula` suffix and use `type_formula` to store the formulas in code. In terms of syntax in the WML, `typed_formula` would mean that unless it's a constant value, you need to enclose the entire formula in parentheses; but it avoids parsing and executing a formula when the value is constant. This also means there'd be (as far as I can tell) no need for two separate damage penalty attributes.

Really good idea. In fact, this is how I wanted to implement things at first, but couldn't find out how to distinguish between formulas and constants.

> Feels like weapon filter and modifications might be missing a few things, though some of them may not have an obvious implementation — filter on hit chance and damage penalties, increase min range, increase (or even transform) hit chance and damage penalties. Perhaps they can wait for later, though (other than increasing min range which seems trivial?).

There are boatloads upon boatloads if these little things that still could be implemented, I tried to draw a line at a sensible place. They can be added later easily if required. Increasing min_range is not difficult (none of these things are), I just didn't think it'd be very useful.

CelticMinstrel commented:
> I'm not sure what do you mean by 'unexpected ways'. It solves a rigorously defined problem. Unless I screwed something up in the implementation, it should give correct results 100% of the time.
>
> I looked up the location filter. It eventually calls get_tiles_in_radius, which uses a different (and I think slightly less efficient) algorithm than the new function. I'm a bit reluctant to ditch a faster implementation for slower one, but probably the correct way to proceed is to refactor my code to use get_tiles_in_radius and then optimize that function later.

I don't have any specific complaints about the algorithm or anything; it's just that if `get_tiles_in_radius` and `get_tiles_in_range` return a different result, it would probably confuse people. In particular, you might be able to run into a situation where some code (maybe AI code) uses a location filter to decide when to attack with a ranged weapon, but then the actual attack fails because the range algorithm was different and the attack's algorithm says you're not actually in range after all. I don't know whether the two algorithms could produce different results, though. It's just something to be careful of.

> There are boatloads upon boatloads if these little things that still could be implemented,

Well, we should at least make sure we don't forget about any of them…

gfgtdf commented:
> I looked up the location filter. It eventually calls get_tiles_in_radius, which uses a different (and I think slightly less efficient) implementation than the new function. I'm a bit reluctant to ditch a faster implementation for slower one, but probably the correct way to proceed is to refactor my code to use get_tiles_in_radius and then optimize that function later.


Probably not, afaik get_tiles_in_radious returns  a vector of locations that is sorted by their distance to the source location.  the later part is probably why its slower than your implementation which seems to require no such order.

CelticMinstrel commented:
If that's the _only_ difference, and the sort order is required in `get_tiles_in_radius`, then the latter could just call the new function and sort it…

gfgtdf commented:
> If that's the only difference, and the sort order is required in get_tiles_in_radius, then the latter could just call the new function and sort it…

that'd definitely be slower than the original implementation.

lipk commented:
> Well, we should at least make sure we don't forget about any of them…

What I was trying to say here that we should be demand-driven when it comes to small niche stuff like this. If it turns out people would like it, we can add them. I can write a list or whatever, but I think if we forget about something that's a pretty good indication that it wasn't that important in the first place.

lipk commented:
> Probably not, afaik get_tiles_in_radious returns a vector of locations that is sorted by their distance to the source location. the later part is probably why its slower than your implementation which seems to require no such order.

That's not what I had in mind. The new algorithm converts stuff to a different coordinate system, where it can efficiently perform the operation, the converts the results back. get_tiles_in_radius works with the original offset coordinates all the way through, that's why I have a hunch that my version is probably faster.

But haven't measured it and there's probably not that much difference. The sorting is a good point, though, I'll keep that in mind.

CelticMinstrel commented:
So do both algorithms guarantee the same result? If they do, then there's no issue as far as I'm concerned.

lipk commented:
Meh, no, we shouldn't have two functions for the same stuff. It adds code, complexity, compile time for nothing. I'll remove mine and maybe look at the existing implementation later.

lipk commented:
I created a project board for the TODOs and further improvements: https://github.com/orgs/wesnoth/projects/1

lipk commented:
@CelticMinstrel I decided not to do the sidebar update thing in the end. The value displayed there always shows the unit's base defense (not accounting for magic, marksman, etc) and I don't thin it's in the scope of this project to change that.

lipk commented:
The other things I wanted to fix have been fixed, although the commits don't show up here for some reason.

gfgtdf commented:
honestly i still don't see the point for hit_chance_penalty and damage_penalty, a [chance_to_hit] (or [damage] ) weapon special can implement this just as well. you can for example do
```
[chance_to_hit]
  name = "Ranged"
  description = "This ranges attacks hit change decreases by 10 percent for every tile of distance betwenn the attacker and the target"
  sub="(10 *  (distance_between(loc, other.loc ) - 1))"
[/chance_to_hit]
```
in your attack cfg file. This will show up in the help just fine, and have the same effect. And is more flexible for example you could also decrease the number of strikes dependent on range via [attacks] or do even facier things like a ranged [drains]  that drains stronger on close units while keeping the damage the same.

This is even more true for a phase were we want authors to experiement and don't really know what will come in the end.

lipk commented:
Sorry I didn't pay attention to this feedback earlier, you make a good point. My WML is very rusty, I didn't realize it was this simple. I feel like some kind of shorthand would be desirable on the long term, but I agree that it's probably a bit premature to add it now, given that the same effect can be reproduced relatively easily.

I'll leave some time for others to chime in on this topic and I'll think about myself, too. If no strong counterargument pops up, I'll remove the penalty attributes. (Not going to move back typed_formula, though, it was very awkwardly placed in GUI2 :))

CelticMinstrel commented:
I don't really mind having two ways of achieving the same thing, as with `hit_chance_penalty` vs a `[chance_to_hit]` special. Such redundancies already exist with the `accuracy` and `parry` attributes. So, I'm not going to step in and say "yes, remove it".

Perhaps an important concern here is how the information can be presented to the player. If having a separate `hit_chance_penalty` attribute enables a more dynamic and useful UI, then I think there's an argument for keeping it as a separate attribute. Or, if you think the penalty should be mixed into the overall hit chance in a way that would differ from a `[chance_to_hit]` special, then that might be another argument. I would tend to lean towards leaving it in just because it was already implemented, as well, though I won't try to say that that's a good argument.

> (Not going to move back typed_formula, though, it was very awkwardly placed in GUI2 :))

I do agree there, and was actually just thinking the same thing myself, that it should perhaps be moved into src/formula. As long as it doesn't rely on anthing in GUI2, but I doubt it does? Maybe some explicit specializations of typed_formula don't belong in src/formula but I guess that can be worked out some other time.

lipk commented:
> Perhaps an important concern here is how the information can be presented to the player. If having a separate hit_chance_penalty attribute enables a more dynamic and useful UI, then I think there's an argument for keeping it as a separate attribute.

I agree with this, however, the PR does not offer a better UI and I'm not sure how that would look like. Currently there's an extra 'Damage' and 'Hit chance' column for the weapon stats table in the unit help page for ranged attacks, which either displays 'n/hex' if the penalty is a constant or 'custom' if it's a formula. In the latter case, we depend on the author properly explaining the penalty somewhere else in the description already. It's rather clumsy in my opinion, but that's the best I could come up with.

> Maybe some explicit specializations of typed_formula don't belong in src/formula but I guess that can be worked out some other time.

Yes, that's what I thought, too. There's currently a GUI2-specific header included solely for the sake of a specialization and I don't like that, but I'm a bit uncertain as to how people would feel about having separate specialization all over the code. That can be very confusing when you're trying to figure out where stuff comes from.

CelticMinstrel commented:
> Currently there's an extra 'Damage' and 'Hit chance' column for the weapon stats table in the unit help page for ranged attacks, which either displays 'n/hex' if the penalty is a constant or 'custom' if it's a formula.

Hmm, the ability to have an extra column seems like a point in favour of keeping this as a separate attribute, but the need to show 'custom' is a downside… and dumping the entire formula into the column probably wouldn't be great either as it could be very long.

> I'm a bit uncertain as to how people would feel about having separate specialization all over the code.

It's definitely not good to have a specialization in its own header… it should either be in the same header as the template, _or_ in the same header as the template argument type that it's specialized on. If the latter is not possible and the former is unwanted, and it's only used in one place, there's also the option of putting it in the one header where it's used.

lipk commented:
> Hmm, the ability to have an extra column seems like a point in favour of keeping this as a separate attribute
It's an okay solution for what I think is going to be the majority of future applications (simple linear penalties), but it's not a huge plus by any means. Actually, I'd be quite happy to kick this can down the road and figure out the UI later, after having finished the fun parts :)

> and dumping the entire formula into the column probably wouldn't be great either as it could be very long.
And not necessarily comprehensible for someone who's not a programmer. I mean, even trivial stuff like multiplication and powers have different notation than what most people are used to.

CelticMinstrel commented:
> And not necessarily comprehensible for someone who's not a programmer. I mean, even trivial stuff like multiplication and powers have different notation than what most people are used to.

Indeed, that is another reason why dumping the formula wouldn't be a great idea.

gfgtdf commented:
> Such redundancies already exist with the accuracy and parry attributes.

This s not really a good reason to keep them, making a mistake once doesn't mean we have to do it again, the accuray and parry specials are afaik not use at all in umc, and in particular not in the way they were meant to be used (making an era with a lot of these attacks to justifie their special handling over [chance_to_hit] specials), so currently all they do is making the wesnoth code larger and probably inducing a bit work for translators who also have to translate related strings.

CelticMinstrel commented:
> This s not really a good reason to keep them

It's not specifically a reason to keep them so much as an observation that similar cases already exist, which in my opinion reduces the motivation for removing the `hit_chance_penalty` attribute.

Here's another possible argument. Do you expect that _most_ people using true ranged attacks will want a penalty? If the penalty is seen as a commonly-desired feature of ranged attacks, then it's more convenient to have a separate attribute rather than forcing (nearly) all ranged attacks to have the same weapon special, which then requires you to decide on a name and description for the weapon special.

lipk commented:
So, I've done some more research and changed my mind again. Implementing penalties with weapon specials doesn't seem to be that straightforward. According to my experiments - and understanding of the code - the weapon special formula can only access the target unit, so distance calculation is actually not possible. Of course this can - and I think should - be rectified with some work, but I'd rather not get sidetracked into it right now.

So the dedicated penalty attributes are here to stay (unless someone steps up and fixes the weapons specials formulas).

I removed the help page interface though. I discovered that the layout falls apart for certain units and it was quite bad even when it didn't.

johncburr commented:
UNSUBSCRIBE

On Sun, Apr 19, 2020 at 11:47 AM lipk <notifications@github.com> wrote:

> Reintroduce real ranged attacks to Wesnoth.
>
> What the PR does:
>
>    - Allows targeting enemies from multiple tiles away
>    - Applies distance-based penalties to hit chance and damage of a weapon
>    - Exposes penalties and ranges to WML and Lua
>    - Displays information about ranged attacks on the unit's help page
>
> Necessary actions after this PR is merged:
>
>    - Update UnitTypeWML docs with new weapon attributes
>    - Update FilterWML docs with new weapon filters
>    - Update EffectWML docs with new weapon modifiers
>
> I believe this is a viable minimal implementation and can be merged back
> to allow content creators to experiment. Nonetheless, there are a couple
> more things to do before this feature can be deemed ready.
>
> TODO:
>
>    - Proper animation support for multihex attacks
>    - Make the AI aware of real ranged attacks
>
> Things that may need further consideration:
>
>    - How the new information is displayed on the help page
>    - Visual aid to indicate which units will be in range when moving to a
>    certain tile
>    - Different attack marker for ranged attacks
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/wesnoth/wesnoth/pull/4859
> Commit Summary
>
>    - Add function to find all tiles in some range n.
>    - Highlight potential ranged targets.
>    - Allow targeting enemies from multiple tiles away
>    - Proper min_range support for real ranged attacks.
>    - Whitespace fixes.
>    - Merge remote-tracking branch 'origin/master' into real-ranged-attacks
>    - Whitespace fix.
>    - Configurable CTH/damage penalties.
>    - Add new attributes to attack_type::write
>    - Ranged attack info in help.
>    - Make new attributes available for weapon mods and lua.
>    - Formatting fixes.
>    - Refactor some really stupid code.
>    - Merge remote-tracking branch 'origin/master' into real-ranged-attacks
>
> File Changes
>
>    - *M* src/actions/attack.cpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-54411f8c204f7ceac25ecc0882ee5623>
>    (9)
>    - *M* src/help/help_topic_generators.cpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-f996419c63b75999b83027d24c62abcf>
>    (25)
>    - *M* src/map/location.cpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-16ce5fb57534ee6b01a3a0123e48fc08>
>    (30)
>    - *M* src/map/location.hpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-56c1c324fd2768511a1dee352bca7e16>
>    (5)
>    - *M* src/mouse_events.cpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-be4b2438e0b85c20fd237f42918b461c>
>    (39)
>    - *M* src/scripting/lua_unit_attacks.cpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-3d96eac3177ac3bb7a6938aa77cdaec0>
>    (12)
>    - *M* src/units/attack_type.cpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-a2e76bb492cc0fe5d8b7228b6bd0ac69>
>    (222)
>    - *M* src/units/attack_type.hpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-9fbd1da2159ac4e69d05a6d63e4de62f>
>    (27)
>    - *M* src/units/unit.cpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-8dd74f653e8c20d19565e37121f74e9f>
>    (18)
>    - *M* src/units/unit.hpp
>    <https://github.com/wesnoth/wesnoth/pull/4859/files#diff-e34b16647c2cbc00e569632821233b3b>
>    (12)
>
> Patch Links:
>
>    - https://github.com/wesnoth/wesnoth/pull/4859.patch
>    - https://github.com/wesnoth/wesnoth/pull/4859.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wesnoth/wesnoth/pull/4859>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFV4OVTHYTKVAEMTG7KHCLRNMTITANCNFSM4ML3QOOA>
> .
>

gfgtdf commented:
> So, I've done some more research and changed my mind again. Implementing penalties with weapon specials doesn't seem to be that straightforward

Soo you are saying that the way i described it above doesn't work?

lipk commented:
Yes. loc and other.loc refer to the same location.

gfgtdf commented:
hmm i was sure we implemented this..

in any case i just made a commit to fix this https://github.com/gfgtdf/wesnoth/commit/b64a7872fdb2a70aa78df35b2f25075940a04b62 is adds "attacker" and "defender" variables which also make it easier to know what  the variables are (instead of 'other'), its not yet om master though

gfgtdf commented:
actuall i just merged the commit in question so it should now be possible,  to do this via 
```
[chance_to_hit]
  name = "Ranged"
  description = "This ranges attacks hit change decreases by 10 percent for every tile of distance betwenn the attacker and the target"
  sub="(10 *  (distance_between(attacker.loc, defender.loc ) - 1))"
[/chance_to_hit]
```

if not then it's a bug in my code (which is not that unlikely since i only tested very briefly yet)

CelticMinstrel commented:
> loc and other.loc refer to the same location.

This sounds like a bug? The backstab special at least relies on `self` and `other` being correctly set (though it doesn't actually access `other.loc`, only `other.facing`). Backstab however uses a formula in the filter, not in the ability, so perhaps _that_ is broken.

> adds "attacker" and "defender" variables which also make it easier to know what the variables are (instead of 'other'), its not yet om master though

Just noting that these new variables can't possibly replace the old `self` and `other` variables, because they're based on completely different criteria:

* `self` is the unit that possesses the ability
* `other` is the opponent of the unit that possesses the ability
* `attacker` is the aggressor in the current fight; it may be equal to either `self` or `other`
* `defender` is the defender in the current fight; it may be equal to either `self` or `other`

So if there's a bug with `other` being set incorrectly, it still needs a fix. In this example, self/other and attack/defender may be interchangeable pairs, but in another example, they won't be.

gfgtdf commented:
> So if there's a bug with other being set incorrectly, it still needs a fix.

If it doesn't work the way it is intended currently, I'd suggest to add a variable "opponent" instead of keeping "other"
(Ideally we could do the same change to filters, but this would require a bit of refactoring and probably involving creating a fomula-context object that is passed to the filter, maybe even directly a formula-map-callable object.)

CelticMinstrel commented:
> If it doesn't work the way it is intended currently, I'd suggest to add a variable "opponent" instead of keeping "other"

I don't think `other` is necessarily the opponent? So renaming it would probably be even more confusing.

gfgtdf commented:
> I don't think other is necessarily the opponent? So renaming it would probably be even more confusing.

I didn't say renaming, i said adding a variable 'opponent' that always refers to the opponent and get rid of 'other' which might refer to whatever, exactly because i think its often unclear to what 'other' exactly refers to. if we need other variables we can add them aswell.

lipk commented:
Just a little heads up - sorry for dropping out of this, I'm a bit caught up with IRL stuff right now, will get back to it when I have the time.

MlemandPurrs commented:
nice, this would open up new possibilities about utilizing the wesnoth engine for creating games that arent restricted to medieval fantasy combat centered around melee.

CelticMinstrel commented:
> get rid of 'other' which might refer to whatever,

It's pretty well-defined what 'other' refers to. The 'self' variable refers to the unit being filtered, while 'other' is the secondary unit of the same filter. If you use a `[filter_adjacent]`, then inside there 'other' refers to the 'self' of the enclosing filter.

AKryukov92 commented:
I know I'm late for party, but I think there are some things worth consideration.

To simplify things about range and elevation, you can allow shooting only in 6 straight directions: north, north-west, north-west, south, south-east, south-east. It will be much clearer to decide, whether this line contain obstacle or not. It also allows more clear formulas about penalties.
Direction "1 hex north, 1 hex NW" is identical to "1 hex NW, 1 hex north", but path lies on different hexes. If one of these intermediate hexes contain mountain while other contain plains, then it is not clear which terrain affects your shot more.

There is one point which is probably missing in this discussion. Will the ranged attack go through other units?
Some attacks might be like mortars - with high trajectory, while others do direct fire and require clear line of sight. Mortars usually have short range while direct fire can reach further targets.

Direct line of fire will require more thoughts about positioning. Should I keep breach in my defence line to allow shooting?

Straight direction of shots and clear line of sight requirement also reduce amount of potential shooters to specific hex.
If you allow to shoot anywhere, then your infantry will be devastated. Because there are 12 hexes on distance 2 and 18 hexes on distance 3. While in melee you have only 6 potential attackers in worst case. Is this a reason to discussion distance penalty?
If you allow only straight direction, then you have 6 attackers on distance 1, another 6 on distance 2 and another 6 on distance 3.
If you force line of sight then you always have 6 attackers regardless of distance.
Of course these options should be configurable. Though they can be simpler than any formulas about changing accuracy and damage from distance.

Attack without straight direction can be treated as "from above", without projectile from shooter. Like Delfadors lightning from the sky. Hey, he is old man but has highest hp among all units. With true ranged attack his hp can be adjusted to reflect his age. Mortars are arcane missiles of mages or stones from trolls.

AI0867 commented:
> There is one point which is probably missing in this discussion. Will the ranged attack go through other units?
> Some attacks might be like mortars - with high trajectory, while others do direct fire and require clear line of sight. Mortars usually have short range while direct fire can reach further targets.

While mortars are typically short range, the longest range fire is all indirect fire, because curving shots can go further. (gravity can cause the projectiles to drop more without them hitting the ground)

> Direct line of fire will require more thoughts about positioning. Should I keep breach in my defence line to allow shooting?
> 
> Straight direction of shots and clear line of sight requirement also reduce amount of potential shooters to specific hex.
> If you allow to shoot anywhere, then your infantry will be devastated. Because there are 12 hexes on distance 2 and 18 hexes on distance 3. While in melee you have only 6 potential attackers in worst case. Is this a reason to discussion distance penalty?
> If you allow only straight direction, then you have 6 attackers on distance 1, another 6 on distance 2 and another 6 on distance 3.
> If you force line of sight then you always have 6 attackers regardless of distance.
> Of course these options should be configurable. Though they can be simpler than any formulas about changing accuracy and damage from distance.
> 
> Attack without straight direction can be treated as "from above", without projectile from shooter. Like Delfadors lightning from the sky. Hey, he is old man but has highest hp among all units. With true ranged attack his hp can be adjusted to reflect his age. Mortars are arcane missiles of mages or stones from trolls.

This all sounds like interesting game design: what are you trying to simulate/what kind of gameplay are you trying to achieve?
All of this should be doable with `[filter_location]`, or with minor extensions of `[filter_location]`.

AKryukov92 commented:
> This all sounds like interesting game design: what are you trying to simulate/what kind of gameplay are you trying to achieve?

By straight direction of shots I mean something like this:
![image](https://user-images.githubusercontent.com/24266616/112892766-940cf400-90fb-11eb-9d88-a3eacf1cf66f.png)


Imagine that Elvish fighter has ranged attack with range=3 and with straight_shots=true and direct_line=true. On the image, all possible directions for shots are "light".
Fighter stand on 10.5, so with straight_shots he can attack enemies on:
north direction: 10.2, 10.3, 10.4
north-east direction: 11.5, 12.4, 13.4
south-east direction: 11.6, 12.6, 13.7
south direction: 10.6, 10.7, 10.8
south-west direction: 9.6. Notice that other allied unit is blocking this direction, so fighter on 10.5 can't attack 7.7 because of direct_line=true
north-west direction: 9.5, 8.4, 7.4

Pentarctagon commented:
Closing due to lack of response. Feel free to re-open later though.
